feat: support using internal certs for the local auth endpoint

banschikovde opened the pull request:
## Issue
Resolves #1299

---

## Problem
When enabling the `local_auth_endpoint` for a cluster, users could not opt into using the internally generated CA certificate.  
Attempting to retrieve the CA certificate through a `data.rancher2_cluster` data source led to a dependency cycle, and only an external CA could be configured directly.

---

## Solution
- Added a `use_internal_ca_certs` boolean flag inside the `local_auth_endpoint` block, mutually exclusive with `ca_certs`.  
- Introduced logic to fetch the cluster’s internal CA and populate it when `use_internal_ca_certs` is true during create, update, and read operations.  
- Updated diff handling and helper functions so the flag is tracked in state and diffed correctly.  
- Expanded resource documentation and unit tests to cover the new behavior.  

---

## Testing

### Engineering Testing
- Ran `make build` to compile the provider.  
- Verified `go test ./rancher2 -run TestExpandClusterV2LocalAuthEndpoint -count=1 -v` after enabling the new flag.  
- Executed `make test` to confirm all unit tests pass with the new logic.  

### Automated Testing
- Added unit tests for `TestExpandClusterV2LocalAuthEndpoint` and updated existing tests to assert state preservation of `use_internal_ca_certs`.  
- Existing unit test suite validates mutual exclusivity between `ca_certs` and `use_internal_ca_certs`.  

---

## QA Testing Considerations
- Confirm that enabling `use_internal_ca_certs` on a cluster with `local_auth_endpoint` correctly pulls internal CA data without requiring explicit `ca_certs`.  
- Validate upgrade scenarios: clusters configured with `ca_certs` should behave unchanged, while toggling `use_internal_ca_certs` should swap to internal CA seamlessly.  

---

## Regressions Considerations
- Potential regression in cluster diffing logic around `local_auth_endpoint` fields; verify that state refresh and plan outputs remain stable.  
- Low probability, but focus on any workflows combining `use_internal_ca_certs` with other cluster CA options.

banschikovde committed:
feat(cluster_v2): add use_internal_ca_certs option for local_auth_endpoint

banschikovde committed:
fix(cluster_v2): correct use_internal_ca_certs schema and update tests

banschikovde committed:
chore(cluster_v2): run gofmt on local_auth_endpoint files

banschikovde committed:
chore(cluster_v2): decode internal CA certs for local auth endpoint

banschikovde committed:
fix(cluster_v2): mark ca_certs field as computed

matttrach commented:
Hello @banschikovde, thanks for the contribution! sorry for the late response, is this still valuable? If so, please resolve any conflicts, rebase off of main and I will get it reviewed. Thank you!

matttrach commented on rancher2/resource_rancher2_cluster_v2.go:
```diff
@@ -51,12 +51,35 @@ func resourceRancher2ClusterV2() *schema.Resource {
 						d.Clear("local_auth_endpoint")
 					} else {
-						d.SetNew("local_auth_endpoint", flattenClusterV2LocalAuthEndpoint(newConfig))
+						d.SetNew("local_auth_endpoint", newObj)
```
maybe I am missing something, but why are we cutting out the flattener here?

banschikovde replied:
flattenClusterV2LocalAuthEndpoint serializes only API fields (ca_certs, enabled, fqdn) and does not include the Terraform-only use_internal_ca_certs flag. If we used the flattener here in CustomizeDiff, that flag would be stripped from the planned object. By passing newObj directly (which already is []interface{} from d.GetChange()), we preserve it.

jiaqiluo replied:
I suggest adding a wrapper around flattenClusterV2LocalAuthEndpoint that appends the use_internal_ca_certs value. This lets us continue filtering out unknown fields while including the new attribute.

matttrach commented on rancher2/structure_cluster_v2_local_auth_endpoint_test.go:
I really appreciate the test additions, thank you!

matttrach commented:
please rebase and resolve any conflicts?

banschikovde commented:
@matttrach 
Yes, this is still valuable! I've resolved the conflicts and rebased off main

matttrach commented on docs/resources/cluster_v2.md:
```diff
@@ -4,7 +4,11 @@ page_title: "rancher2_cluster_v2 Resource"
 Provides a Rancher v2 Cluster v2 resource. This can be used to create node-driver and custom RKE2 and K3s Clusters for Rancher v2 environments and retrieve their information. 
+=======
+Provides a Rancher v2 Cluster v2 resource. This can be used to create RKE2 and K3s Clusters for Rancher v2 environments and retrieve their information.
+>>>>>>> 7e799591 (feat(cluster_v2): add use_internal_ca_certs option for local_auth_endpoint)
```
It looks like you have unresolved conflicts here.

matttrach commented:
Please resolve the conflicts and Copilot suggestions.

jiaqiluo commented:
Hi @banschikovde, 
To my knowledge, the internally generated CA certificate is generally expected to issue certificates only for internal names, such as node hostnames or cluster-local service names.

In your use case, are you using that CA certificate to issue a certificate for the local authentication endpoint? If so, how did you accomplish that?

Could you provide some details about your use case and the approach you took? Thanks!

banschikovde commented:
> Hi @banschikovde, To my knowledge, the internally generated CA certificate is generally expected to issue certificates only for internal names, such as node hostnames or cluster-local service names.
> 
> In your use case, are you using that CA certificate to issue a certificate for the local authentication endpoint? If so, how did you accomplish that?
> 
> Could you provide some details about your use case and the approach you took? Thanks!

The use_internal_ca_certs option was designed to simplify the configuration of local auth endpoints by automatically using the Rancher management cluster's internal CA certificate instead of requiring users to provide an external CA certificate.
The use case is:
- When setting up local auth endpoint with use_internal_ca_certs: true, the provider retrieves the CA certificate from the Rancher management cluster (v1 cluster) and uses it for the local auth endpoint configuration
- This is useful for environments where the Rancher management cluster's CA is trusted by the clients connecting to the local auth endpoint
- It eliminates the need to manually extract and configure the CA certificate

matttrach commented on docs/resources/cluster_v2.md:
```diff
@@ -4,7 +4,11 @@ page_title: "rancher2_cluster_v2 Resource"
 Provides a Rancher v2 Cluster v2 resource. This can be used to create node-driver and custom RKE2 and K3s Clusters for Rancher v2 environments and retrieve their information. 
+=======
+Provides a Rancher v2 Cluster v2 resource. This can be used to create RKE2 and K3s Clusters for Rancher v2 environments and retrieve their information.
+>>>>>>> 7e799591 (feat(cluster_v2): add use_internal_ca_certs option for local_auth_endpoint)
```
This needs to be resolved before we can merge.

jiaqiluo commented:
> > Hi @banschikovde, To my knowledge, the internally generated CA certificate is generally expected to issue certificates only for internal names, such as node hostnames or cluster-local service names.
> > In your use case, are you using that CA certificate to issue a certificate for the local authentication endpoint? If so, how did you accomplish that?
> > Could you provide some details about your use case and the approach you took? Thanks!
> 
> The use_internal_ca_certs option was designed to simplify the configuration of local auth endpoints by automatically using the Rancher management cluster's internal CA certificate instead of requiring users to provide an external CA certificate. The use case is:
> 
> * When setting up local auth endpoint with use_internal_ca_certs: true, the provider retrieves the CA certificate from the Rancher management cluster (v1 cluster) and uses it for the local auth endpoint configuration
> * This is useful for environments where the Rancher management cluster's CA is trusted by the clients connecting to the local auth endpoint
> * It eliminates the need to manually extract and configure the CA certificate

hi @banschikovde, I understand you're trying to make it easier to configure the Authorized Cluster Endpoint (ACE), but I have concerns that this approach is based on a misunderstanding of what these certificate fields represent.  Let me explain:

The `cluster.CACert` field contains the downstream cluster's self-generated Kubernetes API server CA certificate. It's the RKE2's `server-ca.crt` from `/var/lib/rancher/rke2/server/tls/server-ca.crt` (or the K3s equivalent), and is created by RKE2/K3s when the first control plane node bootstraps, and it signs certificates for internal cluster communication with Subject Alternative Names (SANs) like:
  - kubernetes
  - kubernetes.default.svc.cluster.local
  - Node hostnames (e.g., node-1, 172.16.0.10)
  
But NOT for external FQDNs like cluster.example.com. 

One possible way I found is to add the ACE FQDN to the API server certificate's SANs using RKE2's `tls-san` configuration.  **I wonder how could you retrieve and use the CA to sign certs for the authorized cluster endpoint?** 
 
References:
- https://docs.rke2.io/security/certificates#certificate-authority-ca-certificates 
- https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/manage-clusters/access-clusters/authorized-cluster-endpoint
- [How to configure the Authorized Cluster Endpoint with a Layer 4 Loadbalancer for a Rancher-provisioned RKE2 or K3s cluster](https://support.scc.suse.com/s/kb/How-to-configure-the-Authorized-Cluster-Endpoint-with-a-Layer-4-Loadbalancer-for-a-Rancher-provisioned-RKE2-or-K3s-cluster?language=en_US)

towolf commented:
@jiaqiluo I was looking for this feature and I don't understand your questions.

As a user, we are currently using Let's Encrypt on a load balancer in front of the master nodes, and injected their CA root with this
```
  local_auth_endpoint {
    enabled  = true
    fqdn     = "api.${trim(google_dns_managed_zone.cluster.dns_name, ".")}"
    ca_certs = file("./lets-encrypt-chain.pem")
  }
```

But now we want to stop doing that and just proxy TCP and use the native self-signed CA, which comes from `/var/lib/rancher/rke2/server/tls/server-ca.crt`.

But for that to work, we need the generated kubeconfig that comes from Rancher to include it in `certificate-authority-data`, but it's empty. So, this PR would make that work.

That the endpoint is an FQDN is irrelevant for this. Kubectl will use the included CA to validate irrespective of the hostname.

Would be nice if this were in the next release.

jiaqiluo commented on rancher2/structure_cluster_v2_local_auth_endpoint_test.go:
```diff
@@ -26,6 +28,22 @@ func init() {
+			"enabled":               true,
+			"fqdn":                  "fqdn",
+			"use_internal_ca_certs": true,
+		},
```
I'd suggest removing unused variables.

jiaqiluo commented on rancher2/schema_cluster_v2_rke_config_local_auth_endpoint.go:
```diff
@@ -21,6 +22,11 @@ func clusterV2LocalAuthEndpointFields() map[string]*schema.Schema {
 			Type:     schema.TypeString,
 			Optional: true,
 		},
+		"use_internal_ca_certs": {
```
Add description

```suggestion
		// (Optional) Use the cluster's self-signed CA certificate
		// for the authorized cluster endpoint. Only use this when:
		// - Using a Layer 4 load balancer (TCP passthrough)
		// - The cluster's API server certificate includes the FQDN
		//   in its SANs (configured via tls-san in machine_global_config)
		// Mutually exclusive with ca_certs.
		"use_internal_ca_certs": {
```

jiaqiluo reviewed (changes requested):
Left a few comments. 
Please also address Copilot's comments.

matttrach commented:
This repository requires all commits to be signed with a verified key.
Please see [signing-commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information on how to do this.

banschikovde commented:
Hi @matttrach, thanks for the heads up! All commits in this PR are now signed with a verified SSH key. I've rebased the branch and force-pushed - each commit should show as ✅ Verified.

banschikovde committed:
feat: add use_internal_ca_certs option for local_auth_endpoint in cluster_v2

Add a new use_internal_ca_certs boolean field to the local_auth_endpoint
block of rancher2_cluster_v2. This field allows users to instruct Rancher
to use its internal CA certificate when the cluster is behind a Layer 4
load balancer with TLS termination at the ingress controller level.

The field is mutually exclusive with ca_certs — when use_internal_ca_certs
is true, ca_certs must not be set. When enabled, the provider automatically
fetches the internal CA certificate from the Rancher management API.

Also includes:
- Decode base64-encoded CA certs for consistent comparison
- Add description for use_internal_ca_certs schema field
- Wrap errors with %w in getClusterCACert for better context
- Handle d.Set return value in setClusterV2LocalAuthEndpointInternalFlag
- Update tests to reflect mutual exclusivity of ca_certs and use_internal_ca_certs

towolf commented:
Quick question, will it be possible to apply this new setting to existing clusters to change the endpoint to use the internal cert? Because we would like to change our existing clusters to this.

banschikovde commented:
> Quick question, will it be possible to apply this new setting to existing clusters to change the endpoint to use the internal cert? Because we would like to change our existing clusters to this.

Yes, this works on existing clusters. I built the provider from my PR and applied the new config to a couple of existing clusters - it went through without issues

matttrach commented:
Thank you so much for the extended commit message!
